### PR TITLE
Mac: Add PreDelete Notification

### PR DESCRIPTION
**Pull request by wilbaker**
Part of the work required for #152.

**Changes for PreDelete**

- Added `MessageType_KtoU_NotifyFilePreDelete` and `MessageType_KtoU_NotifyDirectoryPreDelete` (kernel->user) notifications
- Updated VFSForGit's user mode code to handle the PreDelete notification
- Added new `Categories.Mac.M2TODO` functional test categories for tests that are required for M2 but not yet passing
- Enabled additional functional tests for file\folder deletes
- Added PreDelete handling to `MirrorProvider`

**Unrelated Fixes\Changes**
- Updated `Message_Init` to properly set the pid and procname in the `MessageHeader`
- Updated `ParseMessageMemory` to set the path to empty string rather than null (when the message from the kernel has a null path)
- Made log formatting more consistent in `MirrorProvider`
- Updated Kext project to treat warnings as errors

**Mac functional test run with these changes**

```
Test Run Summary
  Overall result: Warning
  Test Count: 156, Passed: 154, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 2
    Skipped Tests - Ignored: 2, Explicit: 0, Other: 0
  Start time: 2018-08-17 15:55:50Z
    End time: 2018-08-17 15:59:53Z
    Duration: 242.238 seconds
```

### CODE REVIEW

File: GVFS/GVFS.FunctionalTests/FileSystemRunners/BashRunner.cs [hunk 1]
```diff
@@ -233,7 +239,14 @@ public override void DeleteFile_FileShouldNotBeFound(string path)
 
         public override void DeleteFile_AccessShouldBeDenied(string path)
         {
-            this.DeleteFile(path).ShouldContain(permissionDeniedMessage);
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/FileSystemRunners/BashRunner.cs]: If we start getting more of these, we could create a `WindowsBashRunner` and a `MacBashRunner`. This seems fine for now though.

File: GVFS/GVFS.FunctionalTests/FileSystemRunners/BashRunner.cs [hunk 2]
```diff
@@ -233,7 +239,14 @@ public override void DeleteFile_FileShouldNotBeFound(string path)
 
         public override void DeleteFile_AccessShouldBeDenied(string path)
         {
-            this.DeleteFile(path).ShouldContain(permissionDeniedMessage);
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                this.DeleteFile(path).ShouldContain(permissionDeniedMessage);
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/FileSystemRunners/BashRunner.cs]: Alternatively, don't bother checking the platform, and just make this `ShouldContainOneOf(both messages)`

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs [hunk 3]
```diff
@@ -51,11 +54,12 @@ public void CreateFileInFolderTest()
 
             this.Enlistment.WaitForBackgroundOperations().ShouldEqual(true, "Background operations failed to complete.");
 
-            GVFSHelpers.ModifiedPathsShouldContain(this.fileSystem, this.Enlistment.DotGVFSRoot, folderName + "/" + Environment.NewLine);
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs]: Consider renaming `ModifiedPathsShouldContain` to `ModifiedPathsShouldContainLine`, and move the newline inside the method. That way every caller doesn't have to know this detaill

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs [hunk 4]
```diff
@@ -153,11 +158,13 @@ public void ReadingFileDoesNotUpdateIndexOrSparseCheckout()
             afterUpdateResult.Output.StartsWith("S ").ShouldEqual(true);
             afterUpdateResult.Output.ShouldContain("ctime: 0:0", "mtime: 0:0", "size: 0\t");
 
-            GVFSHelpers.ModifiedPathsShouldNotContain(this.fileSystem, this.Enlistment.DotGVFSRoot, gitFileToCheck + Environment.NewLine);
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.fileSystem, this.Enlistment.DotGVFSRoot, gitFileToCheck + GVFSHelpers.ModifiedPathsNewLine);
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs]: Same for `ShouldNotContainLine`

File: ProjFS.Mac/PrjFSKext/PrjFSKext.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -196,6 +196,7 @@
 				TargetAttributes = {
 					C6C780AF207FC67200E7E054 = {
 						CreatedOnToolsVersion = 9.3;
+						ProvisioningStyle = Automatic;
```
Comment (wilbaker) [inline ProjFS.Mac/PrjFSKext/PrjFSKext.xcodeproj/project.pbxproj]: @sanoursa @nickgra any ideas what this setting does?  At some point xcode decided to add it to the project.

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitMoveRenameTests.cs [hunk 6]
```diff
@@ -139,7 +145,7 @@ public void GitStatusAfterUnstage()
         public void GitStatusAfterFileDelete()
         {
             string existingFilename = "test.cs";
-            this.Enlistment.GetVirtualPathTo(existingFilename).ShouldBeAFile(this.fileSystem);
+            this.EnsureTestFileExists(existingFilename);
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitMoveRenameTests.cs]: Why did you need to change the behavior here?

File: GVFS/GVFS.FunctionalTests/Tools/GVFSHelpers.cs [hunk 7]
```diff
@@ -12,6 +12,8 @@ namespace GVFS.FunctionalTests.Tools
 {
     public static class GVFSHelpers
     {
+        public const string ModifiedPathsNewLine = "\r\n";
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Tools/GVFSHelpers.cs]: Ideally you can make this private as it's not really interesting to the test logic

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 8]
```diff
@@ -215,6 +216,17 @@ void KauthHandler_HandleKernelMessageResponse(uint64_t messageId, MessageType re
...
+        // The follow are not valid responses to kernel messages
+        case MessageType_Invalid:
+        case MessageType_UtoK_StartVirtualizationInstance:
+        case MessageType_UtoK_StopVirtualizationInstance:
+        case MessageType_KtoU_EnumerateDirectory:
+        case MessageType_KtoU_HydrateFile:
+        case MessageType_KtoU_NotifyFileModified:
+        case MessageType_KtoU_NotifyFilePreDelete:
+        case MessageType_KtoU_NotifyDirectoryPreDelete:
+            break;
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Panic? Trace? Disconnect the provider?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 9]
```diff
@@ -284,10 +296,56 @@ static int HandleVnodeOperation(
...
+                    procname,
+                    &kauthResult,
+                    kauthError))
+            {
+                goto CleanupAndReturn;
+            }
+            
+            // This delete could be part of a rename, and so we must hydrate the file now to
+            // ensure it's hydrated post-rename
+            if (kauthResult == KAUTH_RESULT_DEFER &&
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: How could the result currently be anything but defer?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 10]
```diff
@@ -284,10 +296,56 @@ static int HandleVnodeOperation(
...
+                FileFlagsBitIsSet(currentVnodeFileFlags, FileFlags_IsEmpty))
+            {
+                if (!TrySendRequestAndWaitForResponse(
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Can we avoid duplicating this logic for hydrating a file? It seems like `KAUTH_VNODE_DELETE` should just be another action that can trigger a hydration in the existing code path. Then after that, we could check for `KAUTH_VNODE_DELETE` again and trigger the pre-delete notification.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 11]
```diff
@@ -284,10 +296,56 @@ static int HandleVnodeOperation(
...
+                goto CleanupAndReturn;
+            }
+        }
     }
     else
     {
-        if (ActionBitIsSet(
+        if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+        {
+            if (!TrySendRequestAndWaitForResponse(
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Though I'm curious, why are you triggering pre-delete before hydration? Does the ordering make any difference?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 12]
```diff
@@ -412,13 +470,21 @@ static bool ShouldHandleVnodeOpEvent(
...
     {
-        // This vnode is not part of ANY virtualization root, so exit now before doing any more work.
-        // This gives us a cheap way to avoid adding overhead to IO outside of a virtualization root.
-        
-        *kauthResult = KAUTH_RESULT_DEFER;
-        return false;
+        // TODO(Mac): Determine the overhead of making this check for all actions and
+        // vnodes.  If it's too high, this check can be further reduced to the scenarios
+        // that VFSForGit care about.
+        if (!HasParentInVirtualizationRoot(vnode, context))
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Why did you need this change in behaivor?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 13]
```diff
@@ -457,7 +523,12 @@ static bool ShouldHandleVnodeOpEvent(
         
         // TODO(Mac): why do we need to check rootIndex > 0 here?
         // If root was null, we would have already exited. And if not null, it can't be < 0.
-        if (rootIndex >= 0)
+        //
+        // TODO(Mac): Double check that rules are properly enforced when filedFlaggedAsInVirtualizationRoot
+        // is false.  These checks are currently skipped when filedFlaggedAsInVirtualizationRoot is false to
+        // allow providers to create new files inside of their root before starting the virtualization
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Why do we want to allow that?

File: ProjFS.Mac/PrjFSKext/PrjFSKext.xcodeproj/project.pbxproj (hunk 5 above)
Comment (sanoursa) [reply ProjFS.Mac/PrjFSKext/PrjFSKext.xcodeproj/project.pbxproj]: I was going to ask you the same thing :-)

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 14]
```diff
@@ -284,10 +296,56 @@ static int HandleVnodeOperation(
                 }
             }
         }
+        else if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: One more question - do we need to trigger a recursive hydration of this directory? What's the current behavior if you move a virtual directory? What if you move a directory that is non-empty but has virtual items nested deep within it?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 9 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: With VFSForGit it will always be defer, but if a provider decided to block the delete of an empty placeholder it could be DENY.

(The idea here was to only hydrate if the delete is going to be allowed, let me know what you think)

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitMoveRenameTests.cs (hunk 6 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitMoveRenameTests.cs]: Unfortunately all of the tests in this class were written to be completely dependent on *all* of the tests running in order.

`GitStatusAfterFileDelete` was assuming that all prior tests ran, and one of those tests creates "test.cs".  Because we can't run all of the tests just yet there is currently no "test.cs" at the point this test runs.

With this change you can run this individual test without having to run the tests listed above it.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 8 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: At a minimum we should trace.  

As a side-note, I made this change because after enabling "treat warnings as errors" (`"-Werror"`) we started getting compilation errors here (having unhandled enum values in a switch statement triggers a warning).

File: GVFS/GVFS.FunctionalTests/FileSystemRunners/BashRunner.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/FileSystemRunners/BashRunner.cs]: I'd rather not relax the check when we're running the tests on Windows.  I like the idea of separate Windows\Mac BashRunners, but I also agree at this point we probably don't need both.

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs (hunk 3 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs]: I like that idea, that will also allow for making `ModifiedPathsNewLine` private (as mentioned in a separate comment)

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 11 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: > Does the ordering make any difference?

If the provider decides to block the delete it avoids an unnecessary hydration (I don't think VFSForGit would ever do that though ...)

Assuming the provider *allows* the delete, I don't think the ordering would make a difference.  You could argue that the provider wouldn't get the PreDelete notification if the hydration were to fail, but I'm not sure that's enough of a reason to re-order these.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 10 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: I think we could move the PreDelete notification code down and avoid duplicating the logic.  It's a trade-off between avoiding duplicating the logic and preventing an unnecessary hydration if a provider blocks the delete.

However, blocking the delete of an unhydrated would be quite rare (and it's just a perf tradeoff), and so I'll go with your suggestion as I'd rather keep the kext simpler at this point.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 12 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Without this change VFSForGit would not get PreDelete notifications for the index file as it does not have the `FileFlags_IsInVirtualizationRoot` flag set.

In general the `FileFlags_IsInVirtualizationRoot` will only be set for files inside the root that originated as placeholders.  It will not be set for newly created files or files that were renamed into the root.

Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: (And I decided to walk up the tree looking for the flag as we still want to get out the way as fast as we can for files we don't care about, and so I didn't want to add the overhead of looking up the `VirtualizationRoot*` here)

Comment (sanoursa) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: In the prototype, we detected that a new file was created, and set the `FileFlags_IsInVirtualizationRoot` on it. The point of that flag is to give you one spot to check for this optimization. Though I think @pmj might have some plans to get rid of that flag and just do a vnode lookup instead.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 13 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: The problem we have right now is that `gvfs mount` will always attempt to copy in the hooks.

This code didn't cause a problem before because prior to my change above (to proceed even when `FileFlags_IsInVirtualizationRoot` is not set on the file), the kext would bail out before here because neither the .git folder nor the hooks file have the flag set.

I think that once we have the all of the FileOp\Vnode event handling wired up we'll need to revisit how to handle these offline scenarios.

I've filed a follow up issue (#182) for this

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 11 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: See related comment, I'll refactor this code to avoid duplication (and to hydrate first)

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 9 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: See related comment, I'll refactor this code to avoid duplication (and to hydrate first)

File: ProjFS.Mac/PrjFSKext/PrjFSKext.xcodeproj/project.pbxproj (hunk 5 above)
Comment (nickgra) [reply ProjFS.Mac/PrjFSKext/PrjFSKext.xcodeproj/project.pbxproj]: Did you update your XCode version lately? Apple is reworking some things with signing and provisioning profiles, so this is likely related. We should just take the change now (like we do when VS insists on changing a project metadata file), no point in fighting with the IDE.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 12 above)
Comment (sanoursa) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: If doing that work avoids the need for workarounds that we're going to immediately throw away, it may be worth combining them.

**Comment by wilbaker** [comment]
@nickgra @pmj, after seeing the feedback from @sanoursa (and how things can get cleaned up with handling new file notifications) I've decided to make the changes for new file notifications in this PR as well.

Feel free to wait until those changes are ready before reviewing.

Thanks!

**Comment by nickgra** [comment]
LGTM.

**Comment by wilbaker** [comment]
> I've decided to make the changes for new file notifications in this PR as well.

@sanoursa @nickgra @pmj, after looking into what's needed for new file notifications, I've come to the conclusion that it would be cleaner to leave those changes out of this PR.  Rather than adding new file notifications, I'll update this by removing all of the code that's not strictly needed to get basic PreDelete notifications working (e.g. hydrating on delete, delete notifications for renamed files, etc.)

In the next round of PRs (when adding new file and rename notifications), I can add back support for the more complicated scenarios without the need for the work around currently in this PR.

I'll post another message to this PR when it's ready for another round of review.

Thanks!

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs (hunk 3 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs]: I will move the newline inside the method but leave the method named `ModifiedPathsShouldContain`.  

I found `ModifiedPathsShouldContainLine` looked a bit off as the caller is not passing in lines, it's passing in the paths that ``ModifiedPaths` should (or should not) have.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 14 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: I'll leave this investigation to the PR that adds support for the rename notification.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 11 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Update: For now I've removed all of the "hydrate on delete" code.  Interestingly, it seems like we still get some kauth request that hydrates the file before moving.  I'll test more in this area when adding support for rename notifications.

**Comment by wilbaker** [comment]
> after looking into what's needed for new file notifications, I've come to the conclusion that it would be cleaner to leave those changes out of this PR. Rather than adding new file notifications, I'll update this by removing all of the code that's not strictly needed to get basic PreDelete notifications working (e.g. hydrating on delete, delete notifications for renamed files, etc.)

@sanoursa @nickgra @pmj, this cleanup as been completed and the PR is ready for another look.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 15]
```diff
@@ -284,10 +296,38 @@ static int HandleVnodeOperation(
...
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Will the DELETE bit always be exclusive of all the bits we checked for above? You may need to drop this `else` to avoid missing events.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 16]
```diff
@@ -284,10 +296,38 @@ static int HandleVnodeOperation(
...
+                    currentVnode,
+                    pid,
+                    procname,
+                    &kauthResult,
+                    kauthError))
+            {
+                goto CleanupAndReturn;
+            }
+        }
+        else if (ActionBitIsSet(
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: Same here, probably want to drop the `else`
